orttraining/orttraining/training_ops/cuda/communication/nccl_service.cc

```diff
@@ -0,0 +1,346 @@
+    if (schedule_[i].IsAllTasksEqueued() && !schedule_[i].IsAllTasksFinished()) {
+      return i;
+    }
+  }
+  return -1;
+};
+
+void NcclService::SubmitSendAndWait(void* ptr, size_t size, int peer) {
+  // Wait until NCCL service is launched.
+  WaitForLaunch();
+  auto& profile_context = profile::Context::GetInstance();
+  const auto tag = profile_context.GetThreadTagOrDefault(std::this_thread::get_id());
+
+  // Pointer to enqueued task.
+  const NcclTask* task;
+
+  // Submit task.
+  {
+    std::lock_guard<std::mutex> guard(mutex_);
+    auto& profile_context = profile::Context::GetInstance();
+    const auto tag = profile_context.GetThreadTagOrDefault(std::this_thread::get_id());
+    task = schedule_[time_].EqueueTask(NcclTask::Type::SEND, std::vector<int>{peer}, ptr, size, tag);
+  }
+
+  // Wait for task to be finished.
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    cv_.wait(lock, [&] { return task->is_finished; });
+  }
+};
+
+void NcclService::SubmitRecvAndWait(void* ptr, size_t size, int peer) {
+  // Wait until NCCL service is launched.
+  WaitForLaunch();
+
+  // Pointer to euqueued task.
+  const NcclTask* task;
+  {
+    std::lock_guard<std::mutex> guard(mutex_);
+    auto& profile_context = profile::Context::GetInstance();
+    const auto tag = profile_context.GetThreadTagOrDefault(std::this_thread::get_id());
+    task = schedule_[time_].EqueueTask(NcclTask::Type::RECV, std::vector<int>{peer}, ptr, size, tag);
+  }
+
+  // Wait for task to be finished.
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    cv_.wait(lock, [&] { return task->is_finished; });
+  }
+};
+
+void NcclService::Initialize() {
+  // Here we assume GPU i is assigned to local process i.
+  // TODO: Create a general class to describe for computation topology and unify all similar uses.
+  // Hardware a process can own:
+  //   GPUs
+  //   CPUs
+  //   Other devices
+  int mpi_rank;
+  int mpi_size;
+  MPI_CHECK(MPI_Comm_rank(MPI_COMM_WORLD, &mpi_rank));
+  MPI_CHECK(MPI_Comm_size(MPI_COMM_WORLD, &mpi_size));
+
+  // Set device this NCCL communicator runs on.
+  CUDA_CALL(cudaSetDevice(mpi_rank));
+
+  // Get NCCL unique ID at rank 0 and broadcast it to all others.
+  ncclUniqueId id;
+  if (mpi_rank == 0) NCCL_CALL(ncclGetUniqueId(&id));
+  MPI_CHECK(MPI_Bcast((void*)&id, sizeof(id), MPI_BYTE, 0, MPI_COMM_WORLD));
+  NCCL_CALL(ncclCommInitRank(&comm_, mpi_size, id, mpi_rank));
+}
+
+void NcclService::Launch() {
+  worker_ = std::thread([this]() {
+    {
+      std::lock_guard<std::mutex> guard(mutex_);
+      ORT_ENFORCE(is_planned_, "NCCL service must know its communication plan before launching.");
+      // The NCCL service object can only be launched once because it's a
+      // singlton class.
+      ORT_ENFORCE(!is_running_, "NCCL service cannot be repeatedly launched.");
+
+      // Set this flag so that others will not call this again.
+      is_running_ = true;
+      cv_.notify_all();
+    }
+
+    Initialize();
```

Comment:
>Initialize(); [](start = 4, length = 13)

Should Initialize() happen before we set `is_running_` to true and `notify_all()`?